### PR TITLE
byt5 unicode implementation

### PR DESCRIPTION
**Pull request by mchrzanowski**
This PR adds:

* a straightforward implementation of unicode encoding & decoding from [ByT5](https://arxiv.org/abs/2105.13626)
* a very bare-bones text processor for use with this tokenizer that can support any language.

Usage:

To enable this, add the following lines to your nemo launch script:

```
model.encoder_tokenizer.library='byte-level'
model.decoder_tokenizer.library='byte-level'
```

Things work as-is with tarred datasets, but you will need to re-create your datasets using the `byte-level` tokenizer class.

Currently, the following model, which is 260M params, gets within 1 BLEU (18.9) of my transformer-large run on wmt20 ja-en (19.8):

```
    model.encoder_tokenizer.library='byte-level' \
    model.decoder_tokenizer.library='byte-level' \
    model.encoder.max_sequence_length=1024 \
    model.encoder.hidden_size=1536 \
    model.encoder.inner_size=4096 \
    model.encoder.num_attention_heads=16 \
    model.encoder.attn_layer_dropout=0.1 \
    model.encoder.ffn_dropout=0.1 \
    model.encoder.num_layers=10 \
    model.encoder.pre_ln=true \
    model.decoder.max_sequence_length=1024 \
    model.decoder.hidden_size=1536 \
    model.decoder.inner_size=4096 \
    model.decoder.num_attention_heads=16 \
    model.decoder.attn_layer_dropout=0.1 \
    model.decoder.ffn_dropout=0.1 \
    model.decoder.num_layers=2 \
    model.decoder.pre_ln=true \
    model.src_language=ja \
    model.tgt_language=en \
    model.beam_size=3 \
    model.max_generation_delta=3 \
    model.label_smoothing=0.1 \
    model.optim.lr=4e-4 \
    +model.optim.sched.warmup_steps=35000 \
    ~model.optim.sched.warmup_ratio \
```

### CODE REVIEW

File: nemo/collections/common/tokenizers/bytelevel_tokenizer.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
```
Comment (ericharper) [inline nemo/collections/common/tokenizers/bytelevel_tokenizer.py:1]: 2021

File: nemo/collections/common/tokenizers/en_ja_tokenizers.py [hunk 2]
```diff
@@ -25,33 +25,12 @@ class EnJaProcessor:
...
-        return ' '.join(tokens)
+        return text
 
     def normalize(self, text) -> str:
-        # Normalization doesn't handle Japanese periods correctly;
-        # '。'becomes '.'.
-        if self.lang_id == 'en':
-            return self.normalizer.normalize(text)
-        else:
-            return text
```
Comment (ericharper) [inline nemo/collections/common/tokenizers/en_ja_tokenizers.py:57]: Can we get rid of moses style tokenization here?

**Comment by ericharper** [comment]
Could you add sample usage to the PR readme? 

Do you have any results from experiments comparing this tokenizer to yttm? 

Have you used it with tarred datasets?

File: nemo/collections/common/tokenizers/bytelevel_tokenizer.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+from pathlib import Path
+from nemo.collections.common.tokenizers.tokenizer_spec import TokenizerSpec
+
+__all__ = ['ByteLevelTokenizer']
+
+
+class ByteLevelTokenizer(TokenizerSpec):
+    def __init__(self):
+        self.vocab_size = 259
+        self.special_tokens = [self.bos_id, self.eos_id, self.pad_id]
```
Comment (okuchaiev) [inline nemo/collections/common/tokenizers/bytelevel_tokenizer.py:25]: This seems different from ByT5 design which claim to have [padding, eos, unk]. Also, I am wondering if we should have bert-like cls and msk?

File: nemo/collections/common/tokenizers/bytelevel_tokenizer.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        return self.text_to_ids(text)
+
+    def tokens_to_text(self, tokens):
+        return self.ids_to_text(tokens)
+
+    def text_to_ids(self, text):
+        return list(text.encode('utf-8'))
+
+    def ids_to_text(self, ids):
+        # remove special tokens.
```
Comment (okuchaiev) [inline nemo/collections/common/tokenizers/bytelevel_tokenizer.py:38]: @aklife97 is this consistent with other tokenizers' behavior?

**Review by okuchaiev — APPROVED** [review]
looks good to me. One question on special tokens, please see below.

File: nemo/collections/common/tokenizers/en_ja_tokenizers.py (hunk 2 above)
Comment (mchrzanowski) [reply nemo/collections/common/tokenizers/en_ja_tokenizers.py:57]: i think so. in the example of this comment, i can see the model is able to produce both types of periods now

File: nemo/collections/common/tokenizers/bytelevel_tokenizer.py (hunk 3 above)
Comment (mchrzanowski) [reply nemo/collections/common/tokenizers/bytelevel_tokenizer.py:25]: i can remove it, but i think it will break tokenization as a bos token is assumed to be present

File: nemo/collections/common/tokenizers/en_ja_tokenizers.py (hunk 2 above)
Comment (MaximumEntropy) [reply nemo/collections/common/tokenizers/en_ja_tokenizers.py:57]: To follow up on this, I was wondering if instead of just removing all of this content we could keep it as is to still have the BPE way of doing things. We could then use a separate language identifier like en-byte or ja-byte to use this pre/postprocessing.

File: nemo/collections/common/tokenizers/bytelevel_tokenizer.py (hunk 3 above)
Comment (okuchaiev) [reply nemo/collections/common/tokenizers/bytelevel_tokenizer.py:25]: ok, we need bos for decoder in nmt

File: nemo/collections/common/tokenizers/en_ja_tokenizers.py (hunk 2 above)
Comment (mchrzanowski) [reply nemo/collections/common/tokenizers/en_ja_tokenizers.py]: okay, i made a new processor class when dealing with byte-level encoding to preserve the old, default way of doing things

File: nemo/collections/common/tokenizers/bytelevel_tokenizer.py (hunk 1 above)
Comment (mchrzanowski) [reply nemo/collections/common/tokenizers/bytelevel_tokenizer.py]: done

**Comment by mchrzanowski** [comment]
> Could you add sample usage to the PR readme?
can do!
> 
> Do you have any results from experiments comparing this tokenizer to yttm?
we are within 1 bleu of transformer-large on wmt20 ja-en. the relevant part of my launch config has been included in the readme
> 
> Have you used it with tarred datasets?
yep!

**Comment by mchrzanowski** [comment]
i think i've addressed all comments? everyone happy?

**Comment by mchrzanowski** [comment]
i think i've addressed all comments? everyone happy?

**Review by ericharper — APPROVED** [review]
LGTM! Thanks for the updates!

File: nemo/collections/common/tokenizers/en_ja_tokenizers.py [hunk 5]
```diff
@@ -55,3 +39,19 @@ def normalize(self, text) -> str:
             return self.normalizer.normalize(text)
         else:
             return text
+
+
+class EnJaByteLevelProcessor:
```
Comment (ericharper) [inline nemo/collections/common/tokenizers/en_ja_tokenizers.py:44]: This isn't Ja specific. Can we just call it `ByteLevelProcesser` ?

File: nemo/collections/common/tokenizers/en_ja_tokenizers.py [hunk 6]
```diff
@@ -25,28 +25,12 @@ class EnJaProcessor:
...
-        return self.moses_detokenizer.detokenize(tokens)
+        return ' '.join(tokens)
 
     def tokenize(self, text) -> str:
-        """
-        Tokenizes text using Moses. Returns a string of tokens.
-        """
-        tokens = self.moses_tokenizer.tokenize(text)
-        return ' '.join(tokens)
+        return text
```
Comment (ericharper) [inline nemo/collections/common/tokenizers/en_ja_tokenizers.py:33]: We should revert these changes?

**Review by ericharper — CHANGES_REQUESTED** [review]
Sorry, found a couple things to change. Could you see the new comments?

File: nemo/collections/common/tokenizers/en_ja_tokenizers.py (hunk 5 above)
Comment (MaximumEntropy) [reply nemo/collections/common/tokenizers/en_ja_tokenizers.py:44]: @mchrzanowski I think we should call this a `ByteLevelProcessor` and have it be the Processor of choice whenever the tokenizer_library is byte-level regardless of whether the language pair is English-Japanese.

File: nemo/collections/common/tokenizers/en_ja_tokenizers.py (hunk 6 above)
Comment (MaximumEntropy) [reply nemo/collections/common/tokenizers/en_ja_tokenizers.py:33]: I think it's fine to keep this as is if all of the latest En/Ja models don't use any moses tokenization/normalization.

File: nemo/collections/common/tokenizers/bytelevel_tokenizer.py (hunk 4 above)
Comment (aklife97) [reply nemo/collections/common/tokenizers/bytelevel_tokenizer.py:38]: Yupp, it's consistent!

File: nemo/collections/common/tokenizers/en_ja_tokenizers.py (hunk 5 above)
Comment (mchrzanowski) [reply nemo/collections/common/tokenizers/en_ja_tokenizers.py]: done. re-factored the src & target processor selection

**Review by ericharper — APPROVED** [review]
LGTM!

**Review by MaximumEntropy — APPROVED** [review]
Thanks for making all of the changes :)

File: nemo/collections/common/tokenizers/bytelevel_tokenizers.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+
+class ByteLevelProcessor:
+    """
+    A very basic tokenization and detokenization class for use with byte-level
+    tokenization.
+    """
+
+    def detokenize(self, tokens: List[str]) -> str:
+        return ' '.join(tokens)
```
Comment (aklife97) [inline nemo/collections/common/tokenizers/bytelevel_tokenizers.py:30]: A small question, why are we adding whitespace while detokenizing while we are simply returning the text at the time of tokenization?

**Comment by mchrzanowski** [comment]
I afk right now, but I don’t think tokenize and detokenize are supposed to
undo each other. If you look at the batch translate function in the main
nmt file, it will split the string on white space, and so detokenize needs
to reintroduce them.

**Comment by mchrzanowski** [comment]
Thanks, everyone!
